File: cpoly/cauchy.f90
SUBROUTINE cauchy(nn, pt, q, fn_val)
! CAUCHY COMPUTES A LOWER BOUND ON THE MODULI OF THE ZEROS OF A
! POLYNOMIAL - PT IS THE MODULUS OF THE COEFFICIENTS.

INTEGER, INTENT(IN)     :: nn
REAL (dp), INTENT(OUT)  :: pt(:), q(:), fn_val

REAL (dp) :: x, xm, f, dx, df
INTEGER   :: i, n

pt(nn) = -pt(nn)

! COMPUTE UPPER ESTIMATE OF BOUND.
n = nn - 1
x = EXP((LOG(-pt(nn)) - LOG(pt(1))) / n)
IF (pt(n) /= 0.0_dp) THEN

! IF NEWTON STEP AT THE ORIGIN IS BETTER, USE IT.
  xm = -pt(nn) / pt(n)
  IF (xm < x) x = xm
END IF

! CHOP THE INTERVAL (0,X) UNITL F LE 0.
10 xm = x * .1_dp
f = pt(1)
DO  i = 2, nn
  f = f * xm + pt(i)
END DO
IF (f > 0.0_dp) THEN
  x = xm
  GO TO 10
END IF
dx = x

! DO NEWTON ITERATION UNTIL X CONVERGES TO TWO DECIMAL PLACES.
30 IF (ABS(dx/x) > .005_dp) THEN
  q(1) = pt(1)
  DO  i = 2, nn
    q(i) = q(i-1) * x + pt(i)
  END DO
  f = q(nn)
  df = q(1)
  DO  i = 2, n
    df = df * x + q(i)
  END DO
  dx = f / df
  x = x - dx
  GO TO 30
END IF
fn_val = x

RETURN
END SUBROUTINE cauchy

File: cauchy.js
module.exports = function cauchy (nn, pt, q ) {
  // Cauchy computs a lower bound on the moduli ofthe zeros of a polynomial.
  // pt is the modulus of the coefficients.
  //
  // The lower bound of the modulus of the zeros is given by the roots of the polynomial:
  //
  //   n         n-1
  //  z  + |a | z    + ... + |a   | z - |a |
  //         1                 n-1        n
  //
  var x, xmf, dx, df, i;

  // The sign of the last coefficient is reversed:
  pt[nn-1] = -pt[nn-1]

  // Compute upper estimate of bound:
  n = nn - 1;
  x = Math.exp((Math.log(-pt[nn-1]) - Math.log(pt[0])) / n);

  if( pt[n-1] !== 0 ) {
    xm = - pt[nn-1] / pt[n-1];
    if( xm < x ) {
      x = xm;
    }
  }

  // Chop the interval (0,x) until f <= 0
  while( true ) {
    xm = x * 0.1;
    f = pt[0];
    for(i=1; i<nn; i++) {
      f = f * xm + pt[i];
    }
    if( f > 0 ) {
      x = xm;
    } else {
      break;
    }
  }
  dx = x;

  // Do newton iteration until x converges to two decimal places
  while( Math.abs(dx/x) > 0.005 ) {
    q[0] = pt[0];
    for(i=1; i<nn; i++) {
      q[i] = q[i-1] * x + pt[i];
    }
    f = q[nn-1];
    df = q[0];
    for(i=1; i<n; i++) {
      df = df * x + q[i];
    }
    dx = f / df;
    x -= dx;
  }

  return x;
}
